### DISCUSSION TITLE
Hash Function to facilitate refreshing of Aggregate Tables

### DISCUSSION THREAD

**Discussion by Admolly** [duckdb/duckdb#2961] [Ideas]
I am trying to implement what is, in effect, a materialized vew in DuckDB.  Thus, I have a set of base tables I am joining together to create an aggregate table that consists of 19 categorical variables and a handful of statistics.  I want to be able to refresh this table such that I do not have to compute it from scratch each and every time, but right now I am having trouble mapping the rows from the aggregate table back onto the associated rows from the base table in a way that doesn't involve joining on 19 separate columns.

I thought it would be handy to have a hash function I could use to map the 19 categorical variables onto a single integer that would serve as a surrogate key for the aggregate table, and enable me to join on only a single numeric column rather than the 19 categorical ones.  I don't know if DuckDB intends to add support for materialized views at some point,  but until then I feel like there are bound to be other people running into the same issue.   I have not seen anything mentioned in the documentation for this sort of function, but wanted to start a discussion before opening any issues regarding this.

**Comment by Mytherin** [comment]
Perhaps you could use the `rowid` column for this, that every table implicitly has? e.g.:

```sql
D create table integers as select 1 i from range(5);
D select rowid, i from integers;
┌───────┬───┐
│ rowid │ i │
├───────┼───┤
│ 0     │ 1 │
│ 1     │ 1 │
│ 2     │ 1 │
│ 3     │ 1 │
│ 4     │ 1 │
└───────┴───┘
```

**Reply by Admolly** [reply]
right but then there is no correspondence between the base table and aggregate table.  Aggregate table has a 1-to-many relationship with base.

**Reply by Mytherin** [reply]
In that case I think I am not understanding what you are trying to do. What do you mean with "mapping the rows from the aggregate table back onto the associated rows from the base table"?

**Reply by Admolly** [reply]
Sorry, what I am trying to do is establish a single-column foreign key relationship between an aggregate table, and the table it is based off of.  If the aggregate table has columns: [col1, ..., col19, metric1, metric2, etc.] then the base table has: [user_id, col1, ..., col19, metric1, metric2, etc.].  If row "n" from the aggregate table has some ID "x", then I want the corresponding rows from the base table to also have this same ID "x" as a foreign key to the aggregate table.  I hope that is clear.

**Reply by Admolly** [reply]
Here's a bit more practical of an example to demonstrate what I mean:

Let's say we have the table "users" with the following data:

|    |   user_id | country   | device   |   revenue |        surrogate_key |
|---:|----------:|:----------|:---------|----------:|---------------------:|
|  0 |         1 | USA       | mobile   |        50 |   -44721688550912685 |
|  1 |         2 | USA       | mobile   |        25 |   -44721688550912685 |
|  2 |         3 | UK        | desktop  |        10 | -5725795163627909543 |

We make the following aggregate based off this table:

|    | country   | device   |        surrogate_key |   revenue |
|---:|:----------|:---------|---------------------:|----------:|
|  0 | USA       | mobile   |   -44721688550912685 |        75 |
|  1 | UK        | desktop  | -5725795163627909543 |        10 |

Then we update a row in the users table:

|    |   user_id | country   | device   |   revenue |        surrogate_key |
|---:|----------:|:----------|:---------|----------:|---------------------:|
|  0 |         1 | USA       | mobile   |       100 |   -44721688550912685 |
|  1 |         2 | USA       | mobile   |        25 |   -44721688550912685 |
|  2 |         3 | UK        | desktop  |        10 | -5725795163627909543 |

If we want to re-compute the aggregate, but only for the (aggregated) rows which have changed, with the surrogate key we can do the following:

```
SELECT 
  country, 
  device, 
  surrogate_key, 
  sum(revenue) as revenue 
FROM users
JOIN (SELECT surrogate_key FROM users WHERE user_id = 1)
USING (surrogate_key)
GROUP BY 1, 2, 3
```

Without the surrogate key, you would need to perform the self-join using all the categorical variables that constitue the GROUP BY.

Obviously, this example is a bit contrived, but in reality you can easily have millions of rows with tens of categorical variables.  In this case, I believe the performance gains of using a hashed surrogate key will be significant.

**Reply by Mytherin** [reply]
How would you deal with hash collisions correctly without looking at all the keys in detail again?

**Reply by Admolly** [reply]
I suppose you pick a hashing algorithm where the risk is acceptably low and then accept whatever risk that comes with that.

**Reply by Mytherin** [reply]
Once you get enough unique values the risk is 100%, regardless of which hashing algorithm is used. If you map values from a domain with infinite possible values (e.g. strings) to a domain with N possible values, you must encounter a collision after hashing N+1 values, there is no way around that. You will likely find collisions before that point, however.

More to the point, if you are willing to accept this risk (and have a low enough count of unique values), perhaps you could use the md5 function to do this?

**Reply by Admolly** [reply]
I know it's not a great answer but I think that in practice the risk of a hash collision will be low and can be designed to be as low as neccesary.

For example, using 64-bit hashes a collision is unlikely until we approach 2^32 (~4 billion) distinct hashes.  For DuckDB use cases, I don't think there will be too many cases where we approach that limit.  In any case, it is up to the user to decide if it is indeed worth the risk.

The aggregate table I am dealing with has around 7 million distinct sets of categorical variables, for a base table of around 30 million rows (let's say a 1:4 ratio).  I know I am unlikely to get to the point where I have 16 billion rows in the base table as there aren't enough people on the planet :wink:

**Reply by machow** [reply]
Just to revisit this question:

> How would you deal with hash collisions correctly without looking at all the keys in detail again?

In practice people **do** look at all the keys again. For example, validating via tests (e.g. in dbt) that the hashes (keys) are unique. If the hashes failed that test, you could apply a more rigorous process. 

I think that's the key. You can't guarantee there won't be collisions, but in practice it works fine if you are validating uniqueness assumptions (extremely rare they are violated for most data volumes).

**Comment by arjenpdevries** [comment]
If you think about going this route, you probably want to give https://cyan4973.github.io/xxHash/ a thought?

**Reply by Admolly** [reply]
Thank you Arjen that is super helpful!  I will defnitely have a look at this and see what it can do.

**Reply by SartorialOffense** [reply]
I had something like this almost 20 yr ago, created an identity table with an integer key for each combination of attributes (and it was > 19).  Then I just referenced this key in my tables. You also categorize those attributes into multiple "Dimension" tables and just use them in the 2 "Fact" tables - this is very fast and it reduces the number of indices you need because a full table scan of the Dimension table is essentially instant.

I am sure a Hash will work just fine, but it seems more straightforward to use a more standard approach if possible.

**Comment by Admolly** [comment]
Hi everyone, I know this is an old thread but for those who are interested, I thought I would share something related to this that I found the other day.  DBT allows developers to create a [surrogate key](https://docs.getdbt.com/terms/surrogate-key) based on the hashed string value that is derived from the data itself, so apparently there is a precedent for doing this!  Would be neat if it were exposed via the dbt-duckdb API somehow.  Anyways, that's all for now!
